modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/store/useLiferayStore.es.js

```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+
+export default function useLiferayStore(name, reducer, initialState) {
```

Comment:
(Haven't thought about this much, but) it feels like allowing the hook to take the reducer and initial state is putting the concern of bootstrapping the store inside the component when it actually belongs outside of it.

Contrast that with [Redux's hooks](https://react-redux.js.org/api/hooks#usestore), which expect you to set up the store up-front, outside of your React components, and then the hook serves only to get access to it. (And pre-hooks, it was the same: you would "connect" your React components from outside using a HOC pattern, but the store would always be set up outside.)

Reply:
Yeah I agree, it felt a bit odd this way.

One thought I had too was not even having registered stores and instead just a plain global object to work with. This way we can do away with the reducers and registration of stores.

If things got out of control quickly, we might want a more formal structure, but maybe going for simple is best to start with.

Reply:
Yeah, maybe. I don't like going straight to an over-engineered solution when a simple thing would suffice, but at least for the use case that I did the other POC for, we needed more flexibility. ie. it's not just a theoretical, abstract use case.

I'd describe this as a continuum of solutions for coordinating disparate parts of a DXP instance.

1. `Liferay.on`/`Liferay.fire`.
2. Shared global state via simple, plain object.
3. ...
4. ...
5. ...
6. Multiple stores with registration process.
7. ...
8. ...

Given that we already had a motivating use case to go to "6", I don't know why we would want to roll out a solution at "2". It's much easier to take "6" and use it to provide a simplified usage model (ie. a "2", which you can see the basis of in "default" store in the other POC) than it is to take a "2" and evolve it upwards into a "6" or above.

I don't know if we'd ever need to go to "7", "8" etc, but I put them in there to indicate that it's a possibility.

Reply:
> Given that we already had a motivating use case to go to "6", I don't know why we would want to roll out a solution at "2".

Makes sense. I wonder if we can find a middle ground between the two that will satisfy the needs for `6.`.

I'll investigate this further with some of the needs of other teams in mind.